src/app/qgisapp.cpp

```diff
@@ -8003,6 +8003,33 @@ void QgisApp::saveEdits( QgsMapLayer *layer, bool leaveEditable, bool triggerRep
   if ( vlayer == activeLayer() )
     mSaveRollbackInProgress = true;
 
+  // wait for end of rendering to avoid lock in case of
+  // splilte and gpkg fix #15498
+  // Wait only if the current layer is in the rendering pool
+  if ( mMapCanvas->mapSettings().layers().contains( vlayer->id() ) )
+  {
+    if ( vlayer->dataProvider()->storageType().count( "SQLite" ) ||
+         vlayer->dataProvider()->storageType().count( "GPKG" ) )
+    {
+      // notify wait the rendering end
+      if ( mMapCanvas->isDrawing() )
+      {
+        // notify
+        QgsMessageBarItem *item = new QgsMessageBarItem( tr( "Save edit paused" ),
+            tr( "Waiting for end of rendering" ),
+            QgsMessageBar::WARNING, 0 );
+        messageBar()->pushItem( item );
+        // wait
+        while ( mMapCanvas->isDrawing() )
+        {
+          QgsApplication::instance()->processEvents( QEventLoop::AllEvents, 100 );
```

Comment:
Is this just to _sleep_ for 100 ms, or do the app's events specifically need processed? 

This also worked for me and also did not block the GUI:
```cpp
          QEventLoop loop;
          QTimer t;
          connect( &t, SIGNAL( timeout() ), &loop, SLOT( quit() ) );
          t.start( 100 );
          loop.exec();
```
@elpaso, was your concern the usage of `QgsApplication::instance()->processEvents()`? If so, does this handle it more appropriately, and if not, what do you suggest?

Reply:
is for sleep without blocking, events have to be processed without blocking user interaction with the interface. 
The proposed solution use the common and unique event loop.
I can't say if there are drawback on using QgsApplication::instance()->processEvents.

Reply:
BTW I understand it's better an isolated event loop for an isolated feature than interact with common used event loop possibly interfering with other parts of the code.